File: app/src/main/java/com/nativo/sampleapp/ViewAdapter/ListViewAdapter.java
package com.nativo.sampleapp.ViewAdapter;

import static com.nativo.sampleapp.util.AppConstants.CLICK_OUT_URL;
import static com.nativo.sampleapp.util.AppConstants.SECTION_URL;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.nativo.sampleapp.R;

import net.nativo.sdk.NativoSDK;
import net.nativo.sdk.NtvAdData;
import net.nativo.sdk.NtvSectionAdapter;
import net.nativo.sdk.injectable.NtvInjectable;
import net.nativo.sdk.injectable.NtvInjectableType;

/**
 * Example of Nativo SDK implemented using ListView
 * Ads are placed according to rule in link{@code shouldPlaceAdAtIndex()}.
 * If an ad is not placed(eg no fill scenario) the cell is marked with red
 */
public class ListViewAdapter extends BaseAdapter /*implements NtvSectionAdapter*/ {

    private ViewGroup listView;

    public ListViewAdapter(ViewGroup parent) {
        this.listView = parent;
        //NativoSDK.initSectionWithAdapter(this, SECTION_URL, parent.getContext());
    }

    @Override
    public int getCount() {
        return 20;
    }

    @Override
    public Object getItem(int i) {
        return null;
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        if (view == null) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.publisher_article, viewGroup, false);
        }
        if (NativoSDK.getAdTypeForIndex(SECTION_URL, listView, i).equals(NtvInjectableType.VIDEO)) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.video_layout, viewGroup, false);
        } else if (NativoSDK.getAdTypeForIndex(SECTION_URL, listView, i).equals(NtvInjectableType.NATIVE)) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.native_article, viewGroup, false);
        }

        boolean ad = NativoSDK.placeAdInView(view, listView, SECTION_URL, i, null);
        if (!ad) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.native_article, viewGroup, false);
            bindView(view, i);
        }
        return view;
    }

    private void bindView(View view, int i) {
        if (view != null) {
            if (((ImageView) view.findViewById(R.id.article_image)) != null) {
                ((ImageView) view.findViewById(R.id.article_image)).setImageResource(R.drawable.newsimage);
            }
            if (((ImageView) view.findViewById(R.id.sponsored_ad_indicator)) != null) {
                ((ImageView) view.findViewById(R.id.sponsored_ad_indicator)).setVisibility(View.INVISIBLE);
            }
            if (((TextView) view.findViewById(R.id.article_author)) != null) {
                ((TextView) view.findViewById(R.id.article_author)).setText(R.string.sample_author);
            }
            if (((TextView) view.findViewById(R.id.article_title)) != null) {
                ((TextView) view.findViewById(R.id.article_title)).setText(R.string.sample_title);
            }
            if (((TextView) view.findViewById(R.id.sponsored_tag)) != null) {
                ((TextView) view.findViewById(R.id.sponsored_tag)).setVisibility(View.INVISIBLE);
            }
            if (shouldPlaceNativoAdAtIndex(i)) {
                view.findViewById(R.id.article_constraint_layout).setBackgroundColor(Color.RED);
            } else {
                view.findViewById(R.id.article_constraint_layout).setBackgroundColor(Color.WHITE);
            }

            view.setOnClickListener(onClickListener);
        }
    }

    View.OnClickListener onClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            view.getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(CLICK_OUT_URL)));
        }
    };

    public boolean shouldPlaceNativoAdAtIndex(int i) {
        return i % 3 == 0;
    }
/*
    @Override
    public void needsDisplayLandingPage(String sectionUrl, Intent landingPageIntent) {
        listView.getContext().startActivity(landingPageIntent);
    }

    @Override
    public void needsDisplayClickOutURL(String s, String s1) {
        listView.getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(s1)));
    }

    public void hasbuiltView(View view, NtvInjectable ntvInjectable, NtvAdData ntvAdData) {

    }

    public void onReceiveAd(String section, NtvAdData ntvAdData, Integer index) {
        notifyDataSetChanged();
    }

    public void onFail(String section, Integer index) {
        notifyDataSetChanged();
    }

    @Override
    public Class<NtvInjectable> registerInjectableClassForTemplateType(NtvInjectableType templateType, String sectionUrl, Integer index) {
        return null;
    }

    @Override
    public void needsPlaceAdInView(@NonNull String sectionUrl, int atLocation) {
        notifyDataSetChanged();
    }

    @Override
    public void didPlaceAdInView(@NonNull String sectionUrl, int atLocation, @NonNull NtvInjectable adInjectable) {

    }

    @Override
    public void didFailAd(@NonNull String sectionUrl, int atLocation, Throwable error) {
        notifyDataSetChanged();
    }

    @Override
    public void didPrefetchAd(@NonNull String sectionUrl, boolean didGetAd) {

    }

    @Override
    public void didAssignAdToLocation(@NonNull String sectionUrl, int location) {

    }*/
}

File: app/src/main/java/com/nativo/sampleapp/ViewAdapter/ListViewAdapter.kt
package com.nativo.sampleapp.ViewAdapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import com.nativo.sampleapp.R
import com.nativo.sampleapp.util.AppConstants
import com.nativo.sampleapp.util.AppConstants.NtvTAG
import net.nativo.sdk.utils.NativoLayout
import net.nativo.sdk.NativoSDK
import net.nativo.sdk.NtvAdData
import net.nativo.sdk.NtvSectionAdapter
import net.nativo.sdk.injectable.NtvInjectable
import net.nativo.sdk.injectable.NtvInjectableType

/**
 * Example of Nativo SDK implemented using [ListView]
 * Ads are placed according to rule in link`shouldPlaceAdAtIndex()`.
 * If an ad is not placed(eg no fill scenario) the cell is marked with red
 */
class ListViewAdapter(
    private val context: Context, private val listView: ListView
) : BaseAdapter(), NtvSectionAdapter {

    companion object {
        const val ITEM_COUNT = 40
    }

    private var integerList: MutableList<Int> = ArrayList()
    private var initialNativoRequest = true

    /**
     * Show [AppConstants.CLICK_OUT_URL]
     */
    private var itemClickListener = Runnable {
        context.startActivity(
            Intent(
                Intent.ACTION_VIEW,
                Uri.parse(AppConstants.CLICK_OUT_URL)
            )
        )
    }

    init {
        // Nativo init
        NativoSDK.initSectionWithAdapter(this, AppConstants.SECTION_URL, context)

        for (i in 0..ITEM_COUNT) {
            integerList.add(i)
        }
    }

    override fun getCount(): Int {
        return integerList.size
    }

    override fun getItem(p0: Int): Any? {
        return null
    }

    override fun getItemId(p0: Int): Long {
        return 0
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        val item = integerList[i]
        val convertView: View
        if (view != null) {
            if (shouldShowPlaceAd(item)) {
                convertView = if (view !is NativoLayout) {
                    NativoLayout(context)
                } else {
                    view
                }

                NativoSDK.placeAdInView(convertView, listView, AppConstants.SECTION_URL, i, null)
                return convertView
            } else {
                // Publisher article view
                convertView = if (view is NativoLayout) {
                    LayoutInflater.from(context)
                        .inflate(R.layout.publisher_article, viewGroup, false)
                } else {
                    view
                }
                bindView(convertView, integerList[i])
            }
        } else {
            if (shouldShowPlaceAd(item)) {
                convertView = NativoLayout(context)
                NativoSDK.placeAdInView(convertView, listView, AppConstants.SECTION_URL, i, null)
            } else {
                // Publisher article view
                convertView = LayoutInflater.from(context)
                    .inflate(R.layout.publisher_article, viewGroup, false)
                bindView(convertView, integerList[i])
            }
        }

        return convertView
    }

    @SuppressLint("SetTextI18n")
    private fun bindView(view: View, item: Int) {
        val articleImage: ImageView? = view.findViewById(R.id.article_image)
        val articleTitle: TextView? = view.findViewById(R.id.article_title)
        val articleAuthor: TextView? = view.findViewById(R.id.article_description)
        val articleSponsor: ImageView? = view.findViewById(R.id.sponsored_ad_indicator)

        articleImage?.setImageResource(R.drawable.newsimage)
        articleSponsor?.visibility = View.INVISIBLE
        articleAuthor?.setText(R.string.sample_author)
        articleTitle?.text = context.getString(R.string.sample_title) + " :" + item

        view.setOnClickListener {
            itemClickListener.run()
        }
    }

    private fun shouldShowPlaceAd(item: Int): Boolean {
        return item % 3 == 1
    }

    override fun didReceiveAd(didGetFill: Boolean, inSection: String) {
        Log.d(NtvTAG, "Did receive ad: $didGetFill")

        if (didGetFill && initialNativoRequest) {
            Log.w(NtvTAG, "Needs Reload Everything")
            notifyDataSetChanged()
        }

        initialNativoRequest = false
    }

    override fun didAssignAdToLocation(
        location: Int,
        adData: NtvAdData,
        inSection: String,
        container: ViewGroup
    ) {
        Log.d(NtvTAG, "didAssignAdToLocation: $location")
        notifyDataSetChanged()
    }

    override fun didPlaceAdInView(
        view: View,
        adData: NtvAdData,
        injectable: NtvInjectable,
        atLocation: Int,
        inSection: String,
        container: ViewGroup
    ) {
        Log.d(NtvTAG, "didPlaceAdInView: $atLocation AdData: $adData")
    }

    override fun didFailAd(
        inSection: String,
        atLocation: Int?,
        inView: View?,
        container: ViewGroup?,
        error: Throwable?
    ) {
        Log.d(NtvTAG, "onFail at location: $atLocation Error: $error")
        if (atLocation != null && inView != null) {
            Log.w(NtvTAG, "Removing Nativo Ad!")
            integerList.removeAt(atLocation.toInt())
            notifyDataSetChanged()
        }
    }

    override fun needsDisplayLandingPage(
        landingPageIntent: Intent,
        inSection: String,
        container: ViewGroup
    ) {
        context.startActivity(landingPageIntent)
    }

    override fun needsDisplayClickOutURL(url: String, inSection: String, container: ViewGroup) {
        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    override fun <T : NtvInjectable> registerInjectableClassForTemplateType(
        injectableType: NtvInjectableType,
        atLocation: Int?,
        inSection: String?
    ): Class<T>? {
        return null
    }
}